engine/src/main/java/org/hibernate/search/engine/backend/orchestration/spi/BatchingExecutor.java

```diff
@@ -67,274 +55,149 @@ public BatchingExecutor(String name,
-	}
-
-	private void handleCompletion() {
-		// First, tell the processor that we're done processing.
-		long delay = 0;
-		try {
-			delay = processor.completeOrDelay();
-		}
-		catch (Throwable e) {
-			// This will only happen if there is a bug in this class, but we don't want to fail silently
-			FailureContext.Builder contextBuilder = FailureContext.builder();
-			contextBuilder.throwable( e );
-			contextBuilder.failingOperation( "Calling processor.complete() in executor '" + name + "'" );
-			failureHandler.handle( contextBuilder.build() );
+			/*
+			 * Wait for works to complete before trying to handle the next batch.
+			 * Note: timeout is expected to be handled by the processor
+			 * (the Elasticsearch client adds per-request timeouts, in particular),
+			 * so this "join" will not last forever
+			 */
+			Futures.unwrappedExceptionJoin( batchFuture );
 		}
 
-		if ( delay <= 0 ) {
-			// The processor acknowledged that all outstanding operations (commit, ...) have completed.
-			// Tell callers of getCompletion()
-			CompletableFuture<?> justFinishedQueueFuture = this.completionFuture;
-			completionFuture = null;
-			justFinishedQueueFuture.complete( null );
-		}
-		else {
-			// The processor requested that we wait because some outstanding operations (commit, ...)
-			// need to be performed later.
-			scheduledNextProcessing = scheduledExecutorService.schedule(
-					this::ensureProcessingRunning, delay, TimeUnit.MILLISECONDS
-			);
+		@Override
+		public void complete() {
+			processor.complete();
 		}
 	}
 
-	public interface WorkProcessor {
+	private static final class BatchExecutor implements SingletonTask.Executor {
+		private final ExecutorService delegate;
 
-		/**
-		 * Initializes internal state before works are {@link WorkSet#submit(WorkSet) submitted}.
-		 */
-		void beginBatch();
-
-		/**
-		 * Ensures all works {@link WorkSet#submit(WorkSet) submitted} since the last call to {@link #beginBatch()}
-		 * will actually be executed, along with any finishing task (commit, ...).
-		 *
-		 * @return A future completing when the executor is allowed to start another batch.
-		 */
-		CompletableFuture<?> endBatch();
-
-		/**
-		 * Executes any outstanding operation if possible, or return an estimation of when they can be executed.
-		 * <p>
-		 * Called when the executor considers the work queue complete
-		 * and does not plan on submitting another batch due to work starvation.
-		 *
-		 * @return {@code 0} if there is no outstanding operation, or a positive number of milliseconds
-		 * if there are outstanding operations and {@link #completeOrDelay()}
-		 * must be called again that many milliseconds later.
-		 */
-		long completeOrDelay();
-
-	}
-
-	public interface WorkSet<P extends WorkProcessor> {
-
-		void submitTo(P processor);
-
-		void markAsFailed(Throwable t);
-
-	}
-
-	public enum ProcessingStatus {
-
-		IDLE,
-		RUNNING
+		public BatchExecutor(ExecutorService delegate) {
```

Comment:
>  or even better try to remove all the class hierarchy, exposing directly a > java.util.concurrent.ExecutorService as a field in SingletonTask using a flag to mark the different > behavior for LuceneDelayedCommitExecutor exception case.

A single flag is not enough as I need to get the "time to commit" in `LuceneDelayedCommitExecutor`.

I renamed `SingletonTask.Executor` to `SingletonTask.Scheduler`, since that's more consistent with the wording used in `SingletonTask`. This removes the confusion, which I suppose is enough.